File: frontend/src/service/table/Umbuchung.ts
import { z } from 'zod'

import { DateStringSchema, PositionRefSchema } from '../schemas'
import { PositionSchema } from './Bestellung'

export const BestellungUmbuchenSchema = z.object({
  quellTischId: z.number().int().min(1),
  zielTischId: z.number().int().min(1),
  positionen: PositionRefSchema.array().min(1),
  benutzerKommentar: z.string().max(100),
})
export type BestellungUmbuchen = z.infer<typeof BestellungUmbuchenSchema>

// Historien-Eintrag einer geldneutralen Umbuchung. Er erscheint sowohl auf dem
// Quelltisch (Abgang) als auch auf dem Zieltisch (Zugang); die Richtung folgt aus
// dem Verhältnis von tischId zu quellTischId/zielTischId. Nur der Zugang trägt
// stornier-/umbuchbare Positionen (er bringt die Positionen auf den Tisch).
export const UmbuchungSchema = z.object({
  art: z.literal('umbuchung'),
  id: z.uuid(),
  userId: z.number().int().min(1),
  // Eingefrorener Name des Akteurs, beschriftet den Historien-Eintrag.
  userName: z.string().min(1),
  tischId: z.number().int().min(1),
  quellTischId: z.number().int().min(1),
  zielTischId: z.number().int().min(1),
  positionen: PositionSchema.array().min(1),
  gesamtCents: z.number().int().min(0),
  // kommentar ist der Richtungs-Autotext, benutzerKommentar der optionale freie Text.
  kommentar: z.string().max(100),
  benutzerKommentar: z.string().max(100),
  umgebuchtAm: DateStringSchema,
  // Backend-computed (single source of truth); nur für den Zugang befüllt.
  stornierbarePositionen: PositionSchema.array(),
  umbuchbarePositionen: PositionSchema.array(),
})
export type Umbuchung = z.infer<typeof UmbuchungSchema>

File: backend/domain/kasse/umbuchung.go
package kasse

import (
	"time"

	z "github.com/Oudwins/zog"
)

// Umbuchung ist der Historien-Eintrag einer geldneutralen Umbuchung, wie er auf
// Quell- und Zieltisch erscheint.
type Umbuchung struct {
	ID     string
	UserID int
	// UserName ist der eingefrorene Username des Akteurs aus dem Event-Umschlag.
	UserName     string
	TischID      int
	QuellTischID int
	ZielTischID  int
	Positionen   []Position
	GesamtCents  int
	// Kommentar ist der Richtungs-Autotext, BenutzerKommentar der optionale freie Text.
	Kommentar         string
	BenutzerKommentar string
	UmgebuchtAm       time.Time
}

func (u Umbuchung) IstZugang() bool {
	return u.TischID == u.ZielTischID
}

var umbuchungSchema = z.Struct(z.Shape{
	"ID":                z.String().UUID().Required(),
	"UserID":            z.Int().GTE(1).Required(),
	"UserName":          z.String().Min(1).Required(),
	"TischID":           z.Int().GTE(1).Required(),
	"QuellTischID":      z.Int().GTE(1).Required(),
	"ZielTischID":       z.Int().GTE(1).Required(),
	"Positionen":        z.Slice(positionSchema).Min(1).Required(),
	"GesamtCents":       z.Int().GTE(1).Required(),
	"Kommentar":         z.String().Max(100),
	"BenutzerKommentar": z.String().Max(100),
})
